### PR TITLE
refactor [squash]: remove CreateCustomStudySessionListener.kt

### PR DESCRIPTION
**Pull request by david-allison**
## Purpose / Description

* First step in: https://github.com/ankidroid/Anki-Android/pull/17508#pullrequestreview-2465561993
  * remove `CreateCustomStudySessionListener[.Callback]`

## Approach
See commits

## How Has This Been Tested?

Create a Custom Study from DeckPicker & CongratsPage

Trusting CI for the rest

## Learning (optional, can help others)
We need a better method for `dismissAllDialogFragments`

## Checklist
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Comment by mikehardy** [comment]
Hoping for merge on #17508 which is the "blocked by dependency" PR first, that one's a crashfix lined up for release-2.19 and this one will make that one unpickable :-)

This is all known between David and I, but for anyone else reviewing, thought I'd make it explicit

Cheers!

File: AnkiDroid/src/main/java/com/ichi2/anki/CoroutineHelpers.kt [hunk 1]
```diff
@@ -355,7 +355,7 @@ suspend fun <T> Activity.withProgress(
 }
 
 /** @see withProgress(String, ...) */
-suspend fun <T> Fragment.withProgress(message: String, block: suspend () -> T): T =
+suspend fun <T> Fragment.withProgress(message: String = getString(R.string.dialog_processing), block: suspend () -> T): T =
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/CoroutineHelpers.kt:358]: May not be worth it, but a quick scan shows me two callers are explicitly sending in a message that is now the default, is it worth knocking those back to not providing a string that overrides the default with the default?

```
mike@isabela:~/work/ankidroid/Anki-Android/AnkiDroid/src/main/java (fix-custom-study-dialogs-factory-install-crash) % grep -r withProgress *|grep process
com/ichi2/anki/scheduling/SetDueDateDialog.kt:    val cardsUpdated = withProgress(resources.getString(R.string.dialog_processing)) {
com/ichi2/anki/scheduling/ForgetCardsDialog.kt:    withProgress(resources.getString(R.string.dialog_processing)) {
```

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt [hunk 2]
```diff
@@ -441,7 +441,7 @@ class CustomStudyDialog(private val collection: Collection, private val customSt
         collection.decks.save(dyn)
         requireActivity().launchCatchingTask {
             withProgress {
-                rebuildCram(CreateCustomStudySessionListener(customStudyListener!!))
+                rebuildCram(customStudyListener!!)
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt]: Had a note here with contents: `!!` 🧐 
But in the following commit it goes away and is used null-safe 🎉

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt [hunk 3]
```diff
@@ -443,7 +443,7 @@ class CustomStudyDialog(private val collection: Collection, private val customSt
         collection.decks.save(dyn)
         // launch this in the activity scope, rather than the fragment scope
         requireActivity().launchCatchingTask { rebuildDynamicDeck() }
-        // Hide the dialogs
+        // Hide the dialogs (required due to a DeckPicker issue)
         customStudyListener?.dismissAllDialogFragments()
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt:447]: happy to see via code search that these *all* delegate back up to the activity which is AnkiActivity which is the one I just repaired from crashing while attempting to dismiss after activity teardown started

**Review by mikehardy — APPROVED** [review]
Looks great to me, after merge of the crash-fix PR this will need a mechanical application of same stuff (deletion mostly...) to SinglePageActivity

File: AnkiDroid/src/main/java/com/ichi2/anki/CoroutineHelpers.kt (hunk 1 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/CoroutineHelpers.kt:358]: Hmm... I'll pull this one out into an easy PR. deserves a separate commit

Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/CoroutineHelpers.kt:358]: note: These were with `AnkiActivity`, which already had the overrride

**Comment by david-allison** [comment]
* https://github.com/ankidroid/Anki-Android/labels/Blocked%20by%20dependency: https://github.com/ankidroid/Anki-Android/pull/17513

File: AnkiDroid/src/main/java/com/ichi2/anki/CoroutineHelpers.kt (hunk 1 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/CoroutineHelpers.kt:358]: Resolving this comment as it's handled in

* https://github.com/ankidroid/Anki-Android/pull/17513

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt (hunk 3 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt:447]: Just to note, another PR moves `dismissAllDialogFragments` into `FragmentActivity`

* https://github.com/ankidroid/Anki-Android/pull/17511

**Review by BrayanDSO — APPROVED** [review]

**Pull request merged**